analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/ParameterReassignedConstraintTest.cs

```diff
@@ -0,0 +1,90 @@
+    [DataRow(@"arg = Unknown<object[]>();", "params object[]")]
+    [DataRow(@"arg = Unknown<object>(); arg.ToString();", "object")]
+    [DataRow(@"arg = Unknown<object>(); arg.ToString();", "object")]
+    [DataRow(@"arg = Unknown<object>(); arg = Unknown<object>();", "object")]
+    public void ParameterReassignedConstraint_AfterAssignment(string methodSnippet, string argumentType)
+    {
+        var methodBody = $@"{methodSnippet} Tag(""AfterAssignment"", arg);";
+        var argumentSnippet = $", {argumentType} arg";
+        var validator = SETestContext.CreateCS(methodBody, argumentSnippet, new PublicMethodArgumentsShouldBeCheckedForNull()).Validator;
+        validator.ValidateTag("AfterAssignment", x => x.HasConstraint<ParameterReassignedConstraint>().Should().BeTrue());
+    }
+
+    [TestMethod]
+    public void ParameterReassignedConstraint_InsideCondition()
+    {
+        const string snippet = """
+            if (arg == null)
+            {
+                arg = Unknown<object>();
+                Tag("AfterAssignment", arg);
+            }
+            else
+            {
+                Tag("NoAssignment", arg);
+            }
+            """;
+        var validator = SETestContext.CreateCS(snippet, ", object arg", new PublicMethodArgumentsShouldBeCheckedForNull()).Validator;
+        validator.ValidateTag("AfterAssignment", x => x.HasConstraint<ParameterReassignedConstraint>().Should().BeTrue());
+        validator.ValidateTag("NoAssignment", x => x.HasConstraint<ParameterReassignedConstraint>().Should().BeFalse());
+    }
+
+    [TestMethod]
+    public void ParameterReassignedConstraint_WithMultipleParameters()
+    {
+        const string snippet = """
+            arg1 = Unknown<object>();
+            Tag("AfterAssignmentArg1", arg1);
+            Tag("AfterAssignmentArg2", arg2);
+            """;
+        var validator = SETestContext.CreateCS(snippet, ", object arg1, object arg2", new PublicMethodArgumentsShouldBeCheckedForNull()).Validator;
+        validator.ValidateTag("AfterAssignmentArg1", x => x.HasConstraint<ParameterReassignedConstraint>().Should().BeTrue());
+        validator.ValidateTag("AfterAssignmentArg2", x => x.Should().Match(c => c == null || !c.HasConstraint<ParameterReassignedConstraint>()));
+    }
+
+    [DataTestMethod]
+    [DataRow(@"var local = Unknown<object>(); Tag(""AfterAssignment"", local);")]
+    [DataRow(@"ObjectField = Unknown<object>(); Tag(""AfterAssignment"", ObjectField);")]
+    [DataRow(@"FullProperty = Unknown<object>(); Tag(""AfterAssignment"", FullProperty);")]
+    public void PublicMethodArgumentsShouldBeCheckedForNull_IgnoredAssignments_State_Roslyn_CS(string snippet)
+    {
+        var validator = SETestContext.CreateCS(snippet, new PublicMethodArgumentsShouldBeCheckedForNull()).Validator;
+        validator.ValidateTag("AfterAssignment", x => x.Should().Match(c => c == null || !c.HasConstraint<ParameterReassignedConstraint>()));
+    }
```

Comment:
Add a test where after the assignment of a value to a parameter we do the following operations:
* Assignment to field
* Assignment to FullProperty
* Assignment to AutoProperty
* Invocation of an instance method
* Passing p as ref into some other method

```cs
void M(string s)
{
    s = Unknown<string>();
    this.field = null; // <- This part should be parameterized by the test
    Tag("End", s); // s should have the constraint here
}
```
Background: We reset symbol constraints for some symbols on some operations. We need to make sure the constraint survives such resets.

Reply:
I've added the test cases. One of them is failing: _Passing p as ref into some other method_
I removed it from the UTs (see the [last commit](https://github.com/SonarSource/sonar-dotnet/pull/7044/commits/c3fee5a1fc82f972af6d20c27f88bedb638ee503)).
Should I put it back and investigate or open a new issue for it?

Reply:
Make it a separate test case and assert what it looks like now. Add a comment if it is in an undesired state that needs to be fixed.